cognite/client/utils/_priority_tpe.py

```diff
@@ -54,8 +53,23 @@ def python_exit():
         thread.join()
 
 
-atexit.unregister(_python_exit)
-atexit.register(python_exit)
+if sys.version_info[:2] < (3, 9):
+    import atexit
+
+    atexit.unregister(_python_exit)
+    atexit.register(python_exit)
+else:
```

Comment:
A comment explaining why we need the custom shutdown hook at all would be good.

Reply:
_Do_ we actually need it?

Reply:
I've been down the rabbit hole now and yes, we need it, but also _there is a much better solution_ 😂

Reply:
~~~~~~~~
Edit: I was too fast. Need more time to investigate this 🤔  (base class uses the global variable `_threads_queues`, which makes it not trivial to separate in our tpe subclass)

Reply:
I was the one to add the module btw, saw your comment just now. 

One thing that has bothered me with this approach is that we use the regular TPE extensively in all other parallelized calls than datapoints... and thus unregistering the `atexit`* handler seems plain wrong. The PTPE implementation we used was actually quite intertwined with TPE/threading etc. and was a replacement not really able to co-exist... 

Solution: Implement a PTPE-only `_THREADS_QUEUES` and register a separate, additional `atexit`*-handler for it.

(*`atexit` might be "atexit" module or `threading._register_atexit` depending on PY38 or above)

Reply:
**TL;DR**: Now any usage of `ThreadPoolExecutor` together with `PriorityThreadPoolExecutor` is no longer an issue as both have separate cleanup functions registered, without any overlapping reference containers (for threads).